### DISCUSSION TITLE
`clipboard.providers.wl-copy.enable` does not appear to do anything.

### DISCUSSION THREAD

**Discussion by DrVoidest** [nix-community/nixvim#2021] [Q&A] [answered]
Hey, I am just trying to get my clipboard setup for nvim and the `clipboard.providers.wl-copy.enable = true` does not appear to do anything. I have the following in my config 
`clipboard = {
        register = "unnamedplus";
        providers.wl-copy.enable = true;
      };` any idea what I am doing wrong?

**Comment by MattSturgeon** [comment]
Are you using wayland? (wl-copy is meant for use with wayland).

Do you get any errors? (wl-copy works for me on my config :thinking: )

**Reply by DrVoidest** [reply]
Yep hyprland. All I get is no clipboard providers.

**Reply by MattSturgeon** [reply]
Hm, no mention of wl-copy on [hyprland's wiki](https://wiki.hyprland.org/Useful-Utilities/Clipboard-Managers/). It is in their [FAQ](https://wiki.hyprland.org/FAQ/) though.

As you can see in our [clipboard module](https://github.com/nix-community/nixvim/blob/main/modules/clipboard.nix), all we really do is install the enabled provider packages.

Within vim, does `:!which wl-copy` show any output? That should check if `wl-copy` is on the PATH made available to nvim.

**Reply by DrVoidest** [reply]
I get /nix/store/r767lh7928s3k1xrr3ma0r36apr9aia5-wl-clipboard-2.2.1/bin/wl-copy. Running `:set clipboard` outputs `clipboard=`

**Reply by MattSturgeon** [reply]
If you explicitly copy/paste using the `+` register does it work?

It sounds like wl-copy was installed correctly, but the register option didn't correctly set `clipboard` 🤔 

What does `nixvim-print-init` show?

**Reply by DrVoidest** [reply]
[init.txt](https://github.com/user-attachments/files/16606414/init.txt) After re-enabling the register this line got added `clipboard = "unnamedplus"` Still does not work though

**Reply by DrVoidest** [reply]
Tried under a different de still does not work.

**Reply by DrVoidest** [reply]
Trying a virgin config with just the clipboard enabled does not seem to work

**Reply by MattSturgeon** [reply]
That `init.lua` you linked doesn't have any config related to the clipboard. Are you rebuilding (and switching to) your config when testing out changes?

> If you explicitly copy/paste using the `+` register does it work?

Did you try this? i.e. `"+y` and `"+p`

**Reply by DrVoidest** [reply]
Sorry for the late reply and thank you for all your help trying to get this to work for me. Yeah, the most I could get to appear in the init.lua is the `clipboard = "unnamedplus"`, but even explicitly copy/pasting from the + regiser does not work. I have rebuilt it multiple time with different stuff enabled, both doing a rebuild test and switch.

**Reply by GaetanLepage** [reply]
> Yeah, the most I could get to appear in the init.lua is the `clipboard = "unnamedplus"`

The module is not expected to do more than append this line to your init.lua and to add the `wl-copy` package to the PATH in nvim.
Thus, from what you have tested, it seems that our module behaves correctly. Now, why is that not having effect on your specific system is a mystery...

Can you try the following: In nvim, run `:!wl-copy foobar`.
Then, go somewhere else on your system (in your web browser for instance), and simply paste.
For me, it writes `foobar`.

**Reply by MattSturgeon** [reply]
@GaetanLepage is right, our module is doing what it's supposed to, so something else is weird on your system.

In addition to using `wl-copy` directly via `:!wl-copy foobar` you can also check what's in your `:checkhealth` output.

You can search `/provider\.clipboard`. I get:
```
==============================================================================
provider.clipboard: require("provider.clipboard.health").check()

Clipboard (optional) ~
- OK Clipboard tool found: wl-copy

==============================================================================
```

This should be set by nixvim, but for sanity let's also double check `:verbose version`. For me I get:
```
:verbose version
NVIM v0.10.1
Build type: Release
LuaJIT 2.1.1713773202

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/nix/store/cw0l32s7zs3m0pd59chj1grw39vy29vb-neovim-unwrapped-0.10.1/share/nvim"

Run :checkhealth for more info
Press ENTER or type command to continue
```

**Reply by DrVoidest** [reply, marked as answer]
Thank you both, but I think my system is just messed up at this point. I think I might have found the problem, though I was in a root shell editing system files when trying to get it to work, but didn't try testing it as a user. When in a non-root shell, it works, sorry for wasting y'alls time. I assume for security, it is not able to work in a root shell?

**Reply by MattSturgeon** [reply]
Sounds like maybe `wl-copy` doesn't work well when used as root?

> I assume for security, it is not able to work in a root shell?

Generally, yes. You should do as little as possible as root. It's fairly common for people to run vim as root, but it's probably not the best idea given vim is a large enough piece of software that it could easily have vulnerabilities.

While it's probably overkill, if you were worried about editing root files without running vim as root, you could edit a temporary copy of the file owned by your user, then run `cp` as root when you're done.

**Reply by DrVoidest** [reply]
Ok thank you sir

**Reply by MattSturgeon** [reply]
As a minimum I'd recommend using something like `sudo`, `doas` or the new systemd `run0` tool, instead of entering a root shell.

**Reply by ValentinSchwind** [reply]
I've experienced the same clipboard issue as @DrVoidest  when running `sudo nvim somefile` and trying to copy a selection to the clipboard. You can fix this by running `sudo -E nvim somefile`, as mentioned in a [response](https://github.com/bugaevc/wl-clipboard/issues/96#issuecomment-696117547) to an issue raised in wl-clipboard.

As [mentioned](https://github.com/YaLTeR/wl-clipboard-rs/issues/48#issuecomment-2448985156) in another issue on wl-clipboard-rs, running wayland applications as sudo apparently has some security implications (which i have to admit go a little bit over my head). So you might want to consider the already mentioned `cp` workaround or [another method](https://stackoverflow.com/questions/2600783/how-does-the-vim-write-with-sudo-trick-work) for editing system files without sudo / root.

**Reply by DrVoidest** [reply]
@ValentinSchwind Not sure what is different between the clipboard in the two projects, but I moved over to nvf and it works.

**Reply by khaneliman** [reply]
IIRC nvf uses a different neovim wrapper.
